lib/SILOptimizer/Transforms/CopyForwarding.cpp

```diff
@@ -1324,10 +1325,244 @@ class CopyForwardingPass : public SILFunctionTransform
+///
+/// 1. Read-Only Temp Copy: The copy's destination is an AllocStack that is
+/// only written by the initializing copy.
+///
+/// 2. No instructions between the copy and the destroy of its destination
+/// writes to the source. This is sufficient to prove the copy is unnecesary.
+class TempRValueOptPass : public SILFunctionTransform
+{
+  AliasAnalysis *AA = nullptr;
+
+  static bool collectLoads(SILInstruction *CurrentInst, SILInstruction *addr,
+                           llvm::SmallPtrSetImpl<SILInstruction *> &loadInsts);
 
+  bool checkNoSourceModification(CopyAddrInst *copyInst,
+                       const llvm::SmallPtrSetImpl<SILInstruction *> &useInsts);
+
+  bool tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst);
+
+  void run() override;
 };
+
+/// The main entry point of the pass.
+void TempRValueOptPass::run() {
+  DEBUG(llvm::dbgs() << "Copy Peephole in Func " << getFunction()->getName()
+        << "\n");
+
+  AA = PM->getAnalysis<AliasAnalysis>();
+  bool Changed = false;
+
+  // Find all copy_addr instructions.
+  for (auto &BB : *getFunction()) {
+    auto II = BB.begin();
+    while (II != BB.end()) {
+      CopyAddrInst *CopyInst = dyn_cast<CopyAddrInst>(&*II);
+
+      if (CopyInst) {
+        // In case of success, this may delete instructions, but not the
+        // CopyInst itself.
+        Changed |= tryOptimizeCopyIntoTemp(CopyInst);
+      }
+
+      // Increment the instruction iterator now in case the CopyInst is deleted.
+      ++II;
+
+      if (CopyInst && CopyInst->getSrc() == CopyInst->getDest()) {
+        // This is either the CopyInst which just got optimized or it is a
+        // follow-up from an earlier iteration, where another copy_addr copied
+        // the temporary back to the source location.
+        CopyInst->eraseFromParent();
+      }
+    }
+  }
+
+  if (Changed) {
+    invalidateAnalysis(SILAnalysis::InvalidationKind::Instructions);
+  }
+}
+
+/// Transitively explore all data flow uses of the given \p address until
+/// reaching a load or returning false.
+bool TempRValueOptPass::
+collectLoads(SILInstruction *user, SILInstruction *address,
+             llvm::SmallPtrSetImpl<SILInstruction *> &loadInsts) {
+  // All normal uses (loads) must be in the initialization block.
+  // (The destroy and dealloc are commonly in a different block though.)
+  if (user->getParent() != address->getParent())
+    return false;
+
+  // Only allow uses that cannot destroy their operand. We need to be sure
+  // that replacing all this temporary's uses with the copy source doesn't
+  // destroy the source. This way, we know that the destroy_addr instructions
+  // that we recorded cover all the temporary's lifetime termination points.
+  //
+  // Currently we whitelist address projections and loads.
+  //
+  // TODO: handle non-destructive projections of enums
+  // (unchecked_take_enum_data_addr of Optional is nondestructive.)
+  switch (user->getKind()) {
+    default:
```

Comment:
Also, it may be good to handle unchecked_take_enum_data_addr of Optional (no other enums). Optional is pervasive enough, it would be good to handle it as well.